pkg/diagnose/firewall.go

```diff
@@ -118,3 +135,164 @@ func getActiveGatewayNodeName(clusterInfo *cluster.Info, hostname string, status
 
 	return ""
 }
+
+func getGatewayIP(clusterInfo *cluster.Info, localClusterID string, status reporter.Interface) string {
+	gateways, err := clusterInfo.GetGateways()
+	if err != nil {
+		status.Failure("Error retrieving gateways from cluster %q: %v", clusterInfo.Name, err)
+		return ""
+	}
+
+	if len(gateways) == 0 {
+		status.Failure("There are no gateways detected on cluster %q", clusterInfo.Name)
+		return ""
+	}
+
+	for i := range gateways {
+		gw := &gateways[i]
+		if gw.Status.HAStatus != subv1.HAStatusActive {
+			continue
+		}
+
+		for j := range gw.Status.Connections {
+			conn := &gw.Status.Connections[j]
+			if conn.Endpoint.ClusterID == localClusterID {
+				if conn.UsingIP != "" {
+					return conn.UsingIP
+				}
+
+				if conn.Endpoint.NATEnabled {
+					return conn.Endpoint.PublicIP
+				}
+
+				return conn.Endpoint.PrivateIP
+			}
+		}
+	}
+
+	status.Failure("The gateway on cluster %q does not have an active connection to cluster %q", clusterInfo.Name, localClusterID)
+
+	return ""
+}
+
+func verifyConnectivity(localClusterInfo, remoteClusterInfo *cluster.Info, options FirewallOptions,
+	status reporter.Interface, targetPort TargetPort, message string,
+) bool {
+	mustHaveSubmariner(localClusterInfo)
+	mustHaveSubmariner(remoteClusterInfo)
+
+	status.Start(message)
+	defer status.End()
+
+	singleNode, err := remoteClusterInfo.HasSingleNode()
+	if err != nil {
+		status.Failure(err.Error())
+		return false
+	}
+
+	if singleNode {
+		status.Success(singleNodeMessage)
+		return true
+	}
+
+	localEndpoint, err := localClusterInfo.GetLocalEndpoint()
+	if err != nil {
+		status.Failure("Unable to obtain the local endpoint: %v", err)
+		return false
+	}
+
+	gwNodeName := getActiveGatewayNodeName(localClusterInfo, localEndpoint.Spec.Hostname, status)
+	if gwNodeName == "" {
+		return false
+	}
```

Comment:
`getActiveGatewayNodeName` should return an error as well.

```suggestion
	gwNodeName, err := getActiveGatewayNodeName(localClusterInfo, localEndpoint.Spec.Hostname, status)
	if err != nil {
		return err
	}
```